pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -400,108 +339,79 @@ func getVirtualMachineManagedObjectReference(ctx context.Context, c *govmomi.Cli
-	}
-
-	f.SetDatacenter(dc)
-
-	vmRegex := cfg.Global.WorkingDir + filter
-
-	//TODO: get all vms inside subfolders
-	vms, err := f.VirtualMachineList(ctx, vmRegex)
-	if err != nil {
-		return nil, err
-	}
-
-	var vmRef []types.ManagedObjectReference
-	for _, vm := range vms {
-		vmRef = append(vmRef, vm.Reference())
-	}
-
-	pc := property.DefaultCollector(c.Client)
+// Instances returns an implementation of Instances for vSphere.
+func (vs *VSphere) Instances() (cloudprovider.Instances, bool) {
+	return vs, true
+}
 
-	var vmt []mo.VirtualMachine
-	err = pc.Retrieve(ctx, vmRef, []string{"name", "summary"}, &vmt)
-	if err != nil {
-		return nil, err
-	}
+// NodeAddresses is an implementation of Instances.NodeAddresses.
+func (vs *VSphere) NodeAddresses(nodeName k8stypes.NodeName) ([]v1.NodeAddress, error) {
+	addrs := []v1.NodeAddress{}
 
-	var vmList []string
-	for _, vm := range vmt {
-		if vm.Summary.Runtime.PowerState == ActivePowerState {
-			vmList = append(vmList, vm.Name)
-		} else if vm.Summary.Config.Template == false {
-			glog.Warningf("VM %s, is not in %s state", vm.Name, ActivePowerState)
+	if vs.localInstanceID == nodeNameToVMName(nodeName) {
+		/* Get local IP addresses if node is local node */
+		ifaces, err := net.Interfaces()
+		if err != nil {
+			glog.Errorf("net.Interfaces() failed for NodeAddresses - %v", err)
+			return nil, err
 		}
-	}
-	return vmList, nil
-}
-
-type Instances struct {
-	client          *govmomi.Client
-	cfg             *VSphereConfig
-	localInstanceID string
-}
 
-// Instances returns an implementation of Instances for vSphere.
-func (vs *VSphere) Instances() (cloudprovider.Instances, bool) {
-	// Ensure client is logged in and session is valid
-	err := vSphereLogin(context.TODO(), vs)
-	if err != nil {
-		glog.Errorf("Failed to login into vCenter - %v", err)
-		return nil, false
+		for _, i := range ifaces {
+			localAddrs, err := i.Addrs()
+			if err != nil {
+				glog.Warningf("Failed to extract addresses for NodeAddresses - %v", err)
+			} else {
+				for _, addr := range localAddrs {
+					if ipnet, ok := addr.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
+						if ipnet.IP.To4() != nil {
+							// Filter external IP by MAC address OUIs from vCenter and from ESX
+							var addressType v1.NodeAddressType
+							if strings.HasPrefix(i.HardwareAddr.String(), MAC_OUI_VC) ||
+								strings.HasPrefix(i.HardwareAddr.String(), MAC_OUI_ESX) {
+								addressType = v1.NodeExternalIP
+							} else {
+								addressType = v1.NodeInternalIP
+							}
+							v1.AddToNodeAddresses(&addrs,
+								v1.NodeAddress{
+									Type:    addressType,
+									Address: ipnet.IP.String(),
+								},
+							)
+						}
+					}
+				}
+			}
+		}
+		return addrs, nil
 	}
-	return &Instances{vs.client, vs.cfg, vs.localInstanceID}, true
-}
 
-// List returns names of VMs (inside vm folder) by applying filter and which are currently running.
-func (vs *VSphere) list(filter string) ([]k8stypes.NodeName, error) {
```

Comment:
This is removed from Instance interface. It was there in 1.5 but is removed for 1.6.